jobsdb/unionQuery.go

```diff
@@ -0,0 +1,284 @@
+package jobsdb
+
+import (
+	"database/sql"
+	"fmt"
+	"strings"
+)
+
+const (
+	selectQuery = `SELECT jobs.job_id, jobs.uuid, jobs.user_id, jobs.parameters, jobs.custom_val, jobs.event_payload, jobs.event_count, jobs.created_at, jobs.expire_at, jobs.customer, sum(jobs.event_count) over (order by jobs.job_id asc) as running_event_counts `
+)
+
+//used to get pickup-counts during server start-up
+func (jd *HandleT) GetCustomerCounts(defaultBatchSize int) map[string]int {
+	unionMap := make(map[string]int)
+	//just using first DS here
+	//get all jobs all over DSs and then calculate
+	//use DSListLock also
+
+	// jd.dsMigrationLock.RLock()
+	jd.dsListLock.RLock()
+	// defer jd.dsMigrationLock.RUnlock()
+	defer jd.dsListLock.RUnlock()
+	rows, err := jd.dbHandle.Query(fmt.Sprintf(`select customer, count(job_id) from %s group by customer;`, jd.getDSList(false)[0].JobTable))
+	jd.assertError(err)
+
+	for rows.Next() {
+		var customer string
+		var count int
+		err := rows.Scan(&customer, &count)
+		jd.assertError(err)
+		unionMap[customer] = count
+	}
+
+	return unionMap
+}
+
+//Unprocessed
+
+func (jd *HandleT) getUnprocessedUnionQuerystring(unionMap map[string]int, ds dataSetT, params GetQueryParamsT) string {
+	var queries []string
+
+	for customer, count := range unionMap {
+		//do cache stuff here
+		queries = append(queries, jd.getSingleCustomerUnprocessedQueryString(customer, count, ds, params, true))
+	}
+
+	return `(` + strings.Join(queries, `) UNION (`) + `)`
+}
+
+func (jd *HandleT) getSingleCustomerUnprocessedQueryString(customer string, count int, ds dataSetT, params GetQueryParamsT, order bool) string {
+	customValFilters := params.CustomValFilters
+	parameterFilters := params.ParameterFilters
+	var sqlStatement string
+
+	if useJoinForUnprocessed {
+		// event_count default 1, number of items in payload
+		sqlStatement = fmt.Sprintf(
+			selectQuery+
+				`FROM %[1]s AS jobs `+
+				`LEFT JOIN %[2]s AS job_status ON jobs.job_id=job_status.job_id `+
+				`WHERE job_status.job_id is NULL AND customer='%[3]s'`,
+			ds.JobTable, ds.JobStatusTable, customer)
+	} else {
+		sqlStatement = fmt.Sprintf(
+			selectQuery+
+				` FROM %[1]s AS jobs `+
+				`WHERE customer='%[3]s' AND jobs.job_id NOT IN (SELECT DISTINCT(job_status.job_id) FROM %[2]s AS job_status)`,
+			ds.JobTable, ds.JobStatusTable, customer)
+	}
```

Comment:
I think is a good time to decide to keep one implementation or the other. It will simplify our code and reduce the paths we need to test.

Reply:
Yes, it defaults to true. Never seen otherwise, removing the else branch.